frontend/lib/resolution/scope-types.cpp

```diff
@@ -42,40 +54,56 @@ void OwnedIdsWithName::stringify(std::ostream& ss,
 }
 
 llvm::Optional<BorrowedIdsWithName>
-OwnedIdsWithName::borrow(bool skipPrivateVisibilities,
-                         bool onlyMethodsFields) const {
-  if (BorrowedIdsWithName::isIdVisible(idv_,
-                                       skipPrivateVisibilities,
-                                       onlyMethodsFields)) {
-    return BorrowedIdsWithName(idv_, moreIdvs_.get(),
-                               skipPrivateVisibilities, onlyMethodsFields);
+OwnedIdsWithName::borrow(IdAndVis::SymbolTypeFlags filterFlags) const {
+  // Are all of the filter flags present in flagsOr?
+  // If not, it is not possible for this to match.
+  if ((flagsOr_ & filterFlags) != filterFlags) {
+    return llvm::None;
+  }
+
+  if (BorrowedIdsWithName::isIdVisible(idv_, filterFlags)) {
+    return BorrowedIdsWithName(*this, filterFlags);
   }
   // The first ID isn't visible; are others?
   if (moreIdvs_.get() == nullptr) {
     return llvm::None;
   }
 
+  // Are all of the filter flags present in flagsAnd?
+  // If so, return the borrow
+  if ((flagsAnd_ & filterFlags) == filterFlags) {
+    // filter does not rule out anything in the OwnedIds,
+    // so we can return a match.
+    return BorrowedIdsWithName(*this, filterFlags);
+  }
+
+  // Otherwise, use a loop to decide if we can borrow
   for (auto& idv : *moreIdvs_) {
-    if (!BorrowedIdsWithName::isIdVisible(idv,
-                                          skipPrivateVisibilities,
-                                          onlyMethodsFields))
+    if (!BorrowedIdsWithName::isIdVisible(idv, filterFlags))
       continue;
 
-    // Found a visible ID!
-    return BorrowedIdsWithName(idv, moreIdvs_.get(),
-                               skipPrivateVisibilities, onlyMethodsFields);
+    // Found a visible ID! Return a BorrowedIds referring to the whole thing
+    return BorrowedIdsWithName(*this, filterFlags);
```

Comment:
I think this is a regression. `BorrowedIdsWithName` expects the first element to point to a visible ID (ie., `BorrowedIdsWithName::firstId()` should return an actual, visible element, not the first item in the owned list). However, note that you've removed the argument passing `idv` to `BorrowedIdsWithName`, which means that the constructor will pick the first `ibv` from the `OwnedIdsWithName`, which may or may not match the filter.

Reply:
Yes, you are correct. I was trying to get rid of the pointer argument, because it was causing me problems in earlier efforts that added an argument (because C++ is happy to implicitly convert from `bool` to a pointer). I'll adjust it to pass the `idv` separately like did before my changes here (it will still no longer have the pointer argument).